Sources/SwiftProtobuf/ProtobufTextEncoding.swift

```diff
@@ -0,0 +1,291 @@
+// ProtobufRuntime/Sources/Protobuf/ProtobufTextEncoding.swift - Text format encoding support
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+// -----------------------------------------------------------------------------
+///
+/// Text format serialization engine.
+///
+// -----------------------------------------------------------------------------
+
+import Foundation
+
+public struct ProtobufTextEncodingVisitor: ProtobufVisitor {
+    private var encoder = ProtobufTextEncoder()
+    private var tabLevel: Int = 0
+    public var result: String {return encoder.result}
+    
+    
+    public init() {}
+    
+    public init(message: ProtobufTextMessageBase, tabLevel: Int) throws {
+        self.tabLevel = tabLevel
+        
+        try withAbstractVisitor {(visitor: inout ProtobufVisitor) in
+            try message.traverse(visitor: &visitor)
+        }
+    }
+    
+    mutating public func withAbstractVisitor(clause: (inout ProtobufVisitor) throws -> ()) throws {
+        var visitor: ProtobufVisitor = self
+        try clause(&visitor)
+        encoder.text = (visitor as! ProtobufTextEncodingVisitor).encoder.text
+    }
+    
+    
+    mutating public func visitUnknown(bytes: [UInt8]) {
+        // JSON encoding has no provision for carrying proto2 unknown fields
+        // TODO: Does this apply to the text format, as well?
```

Comment:
Text format supports unknown fields (rather, the printer has a flag that lets you control whether they're printed or not). They get printed with their field number: https://github.com/google/protobuf/blob/master/src/google/protobuf/text_format.cc#L1928